Source/Core/Duality/Utility/Pooling/DerivedObjectsPool.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Duality.Utility.Pooling
+{
+	internal class DerivedObjectsPool<T> where T : class
+	{
+		private List<T> pool = new List<T>();
+		private int poolIndex = 0;
+		private int searchIndex = 0;
+
+		public V GetOne<V>() where V : class, T, new()
+		{
+			V result = null;
+
+			for (int i = this.poolIndex; i < this.pool.Count + this.poolIndex; i++)
+			{
+				this.searchIndex = i % this.pool.Count;
+
+				result = this.pool[this.searchIndex] as V;
+
+				if (result != null)
+				{
+					this.poolIndex = this.searchIndex;
+					this.pool.Remove(result);
+
+					break;
+				}
+			}
+
+			if (result == null)
+				result = new V();
+
+			if(result is IPoolable)
+				(result as IPoolable).OnPickup();
```

Comment:
The compiler would hopefully optimize this, but if you do an as cast then check for null it should be faster than doing is followed by as.

Reply:
You can just do the is check nowadays and put the result in a variable: 'result is IPoolable poolable'

Reply:
Ah, my source is still on C# 4, so I couldn't use either elvis nor the inline is-variable... so is the upgrade to the newer C# features agreed on? @ilexp?

Reply:
We have [recently changed the core C# language version to C# 7.3](https://github.com/AdamsLair/duality/blob/4eb524fe5685213661069c4b1976f8781158057e/Source/Core/Duality/Duality.csproj#L30), so both are fine. I personally prefer to stick to the traditional subset of C#, but I wouldn't reject a PR for it. I might occasionally request changes in specific spots, depending on the case.

For now, please avoid elvis though - I have the suspicion that this particular syntactic sugar masks underlying issues with the code that would end up better being fixed at their root.

Reply:
All done